lib/plug/request_id.ex

```diff
@@ -0,0 +1,54 @@
+defmodule Plug.RequestId do
+  @moduledoc """
+  A plug for adding a unique request id to both the request and response headers.
+
+  To use it, just plug it into the desired module and the HTTP
+  request and responses will contain an `x-request-id` header value in
+  the format `uq8hs30oafhj5vve8ji5pmp7mtopc08f`.
+
+      plug Plug.RequestId
+
+  ## Options
+
+    * `:http_header` - The name of the HTTP header to set with a request id. If
+      the HTTP header already exists, its value will be retained. Default
+      value is "x-request-id"
+
+      plug Plug.RequestId, http_header: "custom-request-id"
+  """
+
+  alias Plug.Conn
+  @behaviour Plug
+
+  def init(opts) do
+    Keyword.get(opts, :http_header, "x-request-id")
+  end
+
+  def call(conn, req_id_header) do
+    conn
+    |> get_request_id(req_id_header)
+    |> set_request_id(req_id_header)
+  end
+
+  defp get_request_id(conn, header) do
+    case Conn.get_req_header(conn, header) do
+      []      -> {conn, generate_request_id}
+      [val|_] -> {conn, val}
+    end
+  end
+
+  defp set_request_id({conn, request_id}, header) do
+    conn
+    |> set_req_header(header, request_id)
```

Comment:
> I don't think we should set the request header

While I conceptually disagree (a discussion for a later time), I do agree we shouldn't modify existing behavior. Will revert.